### PR TITLE
added ability to join existing list

### LINKED ISSUE #5: 5. As a user, I want to join an existing shopping list so I can share a shopping list with another person.
Tokens can be shared with other users to allow them to co-manage a given list.

Joining an existing list consists of the following:

* User enters an existing token
* Save the token to `localStorage`
* Show the user the existing list

AC:

* When the user does not already have a token in `localStorage`, on the onboarding/home screen, a simple form is displayed that allows the user to enter a token
* Entering the token and hitting submit saves the token to `localStorage`, effectively giving them joint control of the list
* On submit, show an error if the token does not exist
* If they get an error message, allow them to try again or create a new list
Notes:
* As of this story, the welcome/home screen will be feature complete, as reflected in this wireframe: https://github.com/the-collab-lab/smart-shopping-list/blob/main/_resources/wireframes/welcome-screen.png
* For accessibility reasons, use a `<label>` tag for the "share token" field.
* You can assume that users will only be sharing a token if the list has actual items on it. So, let's say I create a New Shopping List, but never add any items to it, and haven't saved that particular token to Firestore, we can assume that's a list that doesn't exist.
* If you would like, you can save the token to Firestore immediately when creating a list, and then a user would be able to share an empty list (see issue 5).

### PR DESCRIPTION
**Pull request by sparsell**
## Description

If a user is logged in (already has a token in their `localStorage`), they are immediately shown the List page and can continue to enter items to the list. 

If a user is not logged in (does not have a token in `localStorage`), they will be shown the option to create a token or manually enter a valid token which they received from another user. 

An incorrect token will return an error as it could not be found in the Firebase database. 

## Related Issue
closes #5 

## Acceptance Criteria

-- When the user does not already have a token in localStorage, on the onboarding/home screen, a simple form is displayed that allows the user to enter a token

-- Entering the token and hitting submit saves the token to localStorage, effectively giving them joint control of the list

On submit, show an error if the token does not exist

If they get an error message, allow them to try again or create a new list
********
Notes:
-- As of this story, the welcome/home screen will be feature complete, as reflected in this wireframe: https://github.com/the-collab-lab/smart-shopping-list/blob/main/_resources/wireframes/welcome-screen.png
-- For accessibility reasons, use a <label> tag for the "share token" field.
-- You can assume that users will only be sharing a token if the list has actual items on it. So, let's say I create a New Shopping List, but never add any items to it, and haven't saved that particular token to Firestore, we can assume that's a list that doesn't exist.

-- If you would like, you can save the token to Firestore immediately when creating a list, and then a user would be able to share an empty list (see issue 5).

## Type of Changes

|     | Type                       |
| --- | -------------------------- |
|    | :bug: Bug fix              |
|  ✓ | :sparkles: New feature     |
|    | :hammer: Refactoring       |
|    | :100: Add tests            |
|    | :link: Update dependencies |
|    | :scroll: Docs              |

## Updates

-- Added functionality to capture manual token input, query Firestore for that token, and return the list items (documents). 

-- Added error logic for scenario when a token is entered but is not valid

### After

https://user-images.githubusercontent.com/54855300/164915203-160bca06-9061-4ee7-98c7-3b447a8ae7d9.mov

## Testing Steps / QA Criteria

1. Start the app `npm start`
2. If you have a token in `localStorage` you should be taken directly to the List page of that token
3. If you do not have a token, you should see a screen with two user options: 
 -- a button to "Create a new list" 
-- or an input field to enter a token and a button to "Join an existing list" 
4. If you enter a valid token, you should be redirected to the List page and can add more items
5. If you enter an invalid token, you will get an error message and the chance to enter a valid token

### CODE REVIEW

**Review by artsycoder533 — COMMENTED** [review]
- The home screen shows a Create a new list button and an input box to join an existing list.  I removed my token from local storage so this value is empty.
- I hit the Create a new list button and im redirected to the List view.  I check local storage and a token has been created.
- I go back to the home screen and type a random token and I see an error message saying that this token does not match an existing list.
- I pasted in my token value from local storage, but when I hit join an existing list, im not redirected, I stay on the home screen and there is an error "FirebaseError: Function collection() cannot be called with an empty path."
- I added an item to the list, but this time when I try to click on Home im automatically redirected right back to the list, the home screen never shows.  I also noticed firebase didnt automatically upload my token after i added an item to the new list, i had to refresh firebase then it showed up. Maybe thats why im getting an error?

Im not sure if im supposed to approve this or not? I watched the demo video and I tried it out a few times, it just doesnt work like that for me.

**Update:**  It works for me if I paste in an existing list from firebase, I was using the token value from my local storage specificallly.

**Review by daniel-sj-hong — COMMENTED** [review]
So, I got rid of my existing token and opened a new localhost and I was immediately on the `Home` view as intended per the QA criteria. 
I entered in random words in the input field for the token and tried to click the button to `Join an existing list` and the error popped up
I clicked on the `Create a new list` button and a new token was generated for me in `localStorage` and then I was immediately taken to the `List` view. 

However, I think(?) I found a bug and was able to replicate the bug. There is no console error so I'm not 100% sure where the bug is located or if this even is considered a bug. 
- When I click on the `Create a new list` button, it takes me the `List` view with a new token in `localStorage` - nice. 
- I click on the `Home` view and I see my token autofilled into the input box - good.
- I click back on `List` view **and then** I click back on the `Home` view and nothing happens. The `Home` view flashes for a second, but overall I can't go back to the Home view. 

Here's a gif of the error I see and again, nothing shows up on the console as an error so I'm not entirely sure where the bug is coming from if this even is an error. If this is the intended functionality - disregard everything I said and same thing as what Natasha said, it works if I paste in an existing token/list from Firebase. 
Animation

File: src/views/Home/Home.jsx [hunk 1]
```diff
@@ -17,13 +31,33 @@ const Home = ({ token, setToken }) => {
...
+    if (querySnapshot.size > 1) {
+      localStorage.setItem('token', joinList);
+      setToken(joinList);
+      navigate(`/list`);
+    }
+    // otherwise set an error in state we can pass to the join list component
+    setFormError(
+      'This token does not match an existing shopping list. Please check your input and try again.',
+    );
+  };
```
Comment (mxmason) [inline src/views/Home/Home.jsx:51]: 🎉 I know this was a hurdle to get working! Great job! I think your error message is descriptive and helpful – it shows empathy for your user!

File: src/views/Home/Home.jsx [hunk 2]
```diff
@@ -1,14 +1,28 @@
...
 
   useEffect(() => {
     const token = localStorage.getItem('token');
     setHasToken(token !== null);
+    const navigateToList = () => navigate(`/list`);
+    if (hasToken) {
+      navigateToList();
+    }
   }, []);
```
Comment (mxmason) [inline src/views/Home/Home.jsx]: 🧹 Just a note for some code-cleanliness: 
- This hook should probably be made aware of its dependencies (right now, I think that's `token` and `navigate`)
- Do we need `navigateToList` or can we just call `navigate` in our condition here?

**Review by mxmason — APPROVED** [review]
Good work getting this done! :shipit:

File: src/views/Home/Home.jsx (hunk 2 above)
Comment (sparsell) [reply src/views/Home/Home.jsx]: Made suggested changes (thank you!).

**Pull request merged**